### PR TITLE
updates to AMSI script

### LINKED ISSUE #652: [Issue] add German localization for IIS HTTP 400 Error
**Describe the issue**
Running the Test-AMSI script on an Exchange server that has a different OS language than English (German in my case) will lead to the warning message that AMSI may be disabled, as the script checks for the English error message of the code 400 / Bad Request that is returned. The test is working fine despite of this, code 400 is returned, and the detection is logged. And a Microsoft 365 Defender alert was triggered, as we've onboarded the Exchange servers to Defender for Endpoint. Of course, you can change the script and add the error message in the OS language if you want it to display the correct result.

**Expected behavior**
A correct localized message

### PR DESCRIPTION
**Pull request by tweekerz**
Fixes #652 #653

### CODE REVIEW

**Review by lusassl-msft — APPROVED** [review]

**Comment by lusassl-msft** [comment]
Looks like there is an formatting issue.

File: Admin/Test-AMSI.ps1 [hunk 1]
```diff
@@ -1,11 +1,9 @@
-﻿# Copyright (c) Microsoft Corporation.
-# Licensed under the MIT License.
```
Comment (bill-long) [inline Admin/Test-AMSI.ps1:2]: Compliance header was removed. It is required.

File: Admin/Test-AMSI.ps1 [hunk 2]
```diff
@@ -1,11 +1,9 @@
...
-
-[CmdletBinding(DefaultParameterSetName = "TestAMSI", HelpUri = "https://aka.ms/css-exchange")]
+﻿[CmdletBinding(DefaultParameterSetName = "TestAMSI", HelpUri = "https://aka.ms/css-exchange")]
 param(
     [Parameter(ParameterSetName = 'TestAMSI', Mandatory = $true, Position = 0)]
     [ValidateNotNullOrEmpty()]
     [string]$ExchangeServerFQDN,
+    [switch]$IgnoreSSL,
```
Comment (bill-long) [inline Admin/Test-AMSI.ps1:6]: Trailing whitespace on line 6.

**Review by bill-long — CHANGES_REQUESTED** [review]

**Comment by bill-long** [comment]
`.build\CodeFormatter.ps1 -Save` will fix both issues.

File: Admin/Test-AMSI.ps1 [hunk 3]
```diff
@@ -37,14 +35,29 @@ function Test-AMSI {
...
+        public class TrustAllCertsPolicy : ICertificatePolicy {
+            public bool CheckValidationResult(
+                ServicePoint srvPoint, X509Certificate certificate,
+                WebRequest request, int certificateProblem) {
+                return true;
+            }
+        }
+"@
+        Add-Type -TypeDefinition $IGSSL
+        [System.Net.ServicePointManager]::CertificatePolicy = New-Object TrustAllCertsPolicy
```
Comment (bill-long) [inline Admin/Test-AMSI.ps1:51]: CertificatePolicy is obsolete. You can delete all this code and add one line instead:

`[Net.ServicePointManager]::ServerCertificateValidationCallback = {$true}`

Comment (lusassl-msft) [reply Admin/Test-AMSI.ps1:51]: Oh, didn’t know that. @tweekerz https://docs.microsoft.com/en-us/dotnet/api/system.net.servicepointmanager.certificatepolicy?view=netframework-4.8 for reference.

Comment (bill-long) [reply Admin/Test-AMSI.ps1:51]: Due to runspace errors when we try to use a PowerShell method for the callback, we'll have to set the ServerCertificateValidationCallback property from C# like so:

```
Add-Type @"
    using System;
    using System.Net;
    using System.Net.Security;
    using System.Security.Cryptography.X509Certificates;
    public class ServerCertificateValidationBehavior
    {
        public static void Ignore()
        {
            ServicePointManager.ServerCertificateValidationCallback += 
                delegate(Object obj, X509Certificate certificate, X509Chain chain, SslPolicyErrors errors)
                {
                    return true;
                };
        }
    }
"@

[ServerCertificateValidationBehavior]::Ignore()
```

We should also remove this when we're done by setting it back to null. This, we can do from PowerShell:

```
[Net.ServicePointManager]::ServerCertificateValidationCallback = $null
```

```

**Review by bill-long — APPROVED** [review]

**Comment by bill-long** [comment]
One oddity to note is that the shell continues to ignore certificate errors for some time after that setting is used. We seem to be caching the Invoke-WebRequest session or something, and I haven't found a way to make the $null immediately take effect. It would be nice if we could eventually address this somehow.

image

**Pull request merged**